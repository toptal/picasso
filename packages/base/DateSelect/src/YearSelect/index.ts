export type { Props as YearSelectProps } from './YearSelect'
export { default as YearSelect } from './YearSelect'
