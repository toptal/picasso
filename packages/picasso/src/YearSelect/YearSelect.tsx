import React, { useMemo, Ref } from 'react'

import Select, { SelectProps } from '../Select'
import { forwardRef, documentable } from '../utils/forward-ref'

type AdjustedSelectProps<M extends boolean> = Omit<
  SelectProps<number, M>,
  'options'
>

/**
 * YearSelect props are generalized over multiselect property. If you want `onChange`
 * to take handler that can take array (for multiselect) you should set `Multiple`
 * to `true`. By default it's single select.
 *
 * @param Multiple The `boolean` type of the `multiple` property to indicate whether `onChange` will expect handler to accept plain `number` or array of `number`s
 */
export interface Props<Multiple extends boolean>
  extends AdjustedSelectProps<Multiple> {
  /** a year select starts from. e.g. 2017 */
  from: number
  /** a year select ends at. e.g. 2019 */
  to: number
}

const generateOptions = (from: number, to: number) => {
  const length = Math.abs(to - from) + 1

  return Array.from({ length }).map((_, index) => {
    const value = to < from ? from - index : index + from

    return {
      value,
      text: value.toString(),
    }
  })
}

export const YearSelect = documentable(
  forwardRef(
    <M extends boolean = false>(
      { from, to, ...rest }: Props<M>,
      ref: Ref<HTMLInputElement>
    ) => {
      if (!to || !from) {
        throw new Error(
          `Invalid range. Please check the values you have passed: from: ${from}, to: ${to}`
        )
      }

      const options = useMemo(() => generateOptions(from, to), [from, to])

      return <Select {...rest} ref={ref} options={options} />
    }
  )
)

YearSelect.defaultProps = {}

YearSelect.displayName = 'YearSelect'

export default YearSelect
