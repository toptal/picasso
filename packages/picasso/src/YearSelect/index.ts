export type { Props as YearSelectProps } from './YearSelect'
export { default } from './YearSelect'
