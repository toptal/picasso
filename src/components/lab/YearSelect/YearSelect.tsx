import React, { forwardRef, useMemo, ChangeEvent } from 'react'

import Select, { Props as SelectProps } from '../../Select'
import { JssProps, OmitInternalProps } from '../../Picasso'

type AdjustedSelectProps = OmitInternalProps<
  Omit<SelectProps, 'onChange' | 'options'>
> &
  Partial<JssProps>

export interface Props extends AdjustedSelectProps {
  /** a year select starts from. e.g. 2017 */
  from: number
  /** a year select ends at. e.g. 2019 */
  to: number
  /** Callback invoked when picker changes its state. */
  onChange: (
    event: ChangeEvent<{ name?: string | undefined; value: unknown }>
  ) => void
}

function generateOptions(from: number, to: number) {
  const length = to - from + 1

  return Array.from({ length }).map((_, index) => {
    const value = index + from

    return {
      value,
      text: value.toString()
    }
  })
}

export const YearSelect = forwardRef<HTMLInputElement, Props>(
  function YearSelect({ from, to, onChange, ...rest }, ref) {
    const handleChange = (
      event: ChangeEvent<{ name?: string | undefined; value: unknown }>
    ) => {
      onChange(event)
    }

    if (!to || !from || to < from) {
      throw new Error(
        `Invalid range. Please check the values you have passed: from: ${from}, to: ${to}`
      )
    }

    const options = useMemo(() => generateOptions(from, to), [from, to])

    return (
      // eslint-disable-next-line react/jsx-props-no-spreading
      <Select {...rest} ref={ref} options={options} onChange={handleChange} />
    )
  }
)

YearSelect.defaultProps = {}

YearSelect.displayName = 'YearSelect'

export default YearSelect
