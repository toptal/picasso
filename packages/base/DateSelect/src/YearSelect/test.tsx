import React from 'react'
import { render, fireEvent } from '@toptal/picasso-test-utils'

import { YearSelect } from './YearSelect'

// Mock console error since toThrow() outputs the error message with stacktrace
let mockedConsoleError: jest.SpyInstance

describe('YearSelect', () => {
  beforeEach(() => {
    mockedConsoleError = jest.spyOn(console, 'error')
    mockedConsoleError.mockImplementation(() => {})
  })

  afterEach(() => {
    mockedConsoleError.mockRestore()
  })

  it('renders', () => {
    const { container } = render(
      <YearSelect from={2001} to={2005} onChange={() => {}} />
    )

    expect(container).toMatchSnapshot()
  })

  it('render in descending order', () => {
    const placeholder = 'Select year'
    const { getByPlaceholderText, getByRole } = render(
      <YearSelect
        from={2005}
        to={2001}
        placeholder={placeholder}
        onChange={() => {}}
      />
    )

    fireEvent.click(getByPlaceholderText(placeholder))

    expect(getByRole('listbox')).toMatchSnapshot()
  })

  it('wrong range', () => {
    let from: number | null, to: number | null

    const errorMessage = () =>
      `Invalid range. Please check the values you have passed: from: ${from}, to: ${to}`

    // casting is needed to trick typescript and pass null
    const tryRender = () =>
      render(
        <YearSelect
          onChange={() => {}}
          from={from as number}
          to={to as number}
        />
      )

    from = null
    to = 10
    expect(tryRender).toThrow(errorMessage())

    from = 10
    to = null
    expect(tryRender).toThrow(errorMessage())
  })
})
