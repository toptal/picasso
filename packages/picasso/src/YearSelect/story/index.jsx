import { YearSelect } from '../YearSelect'
import PicassoBook from '~/.storybook/components/PicassoBook'

const page = PicassoBook.section('Forms').createPage(
  'YearSelect',
  `Year select

  ${PicassoBook.createSourceLink(__filename)}
  `
)

page
  .createTabChapter('Props')
  .addComponentDocs({ component: YearSelect, name: 'YearSelect' })

page
  .createChapter()
  .addExample('YearSelect/story/Default.example.tsx', 'Default')
