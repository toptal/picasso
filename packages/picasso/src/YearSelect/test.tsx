import React from 'react'
import { render } from '@testing-library/react'
import Picasso from '@toptal/picasso-shared'

import YearSelect from './YearSelect'

// Mock console error since toThrow() outputs the error message with stacktrace
let mockedConsoleError: jest.SpyInstance

beforeEach(() => {
  mockedConsoleError = jest.spyOn(console, 'error')
  mockedConsoleError.mockImplementation(() => {})
})

afterEach(() => {
  mockedConsoleError.mockRestore()
})

describe('YearSelect', () => {
  test('default render', () => {
    const { container } = render(
      <Picasso loadFonts={false}>
        <YearSelect from={2001} to={2005} onChange={() => {}} />
      </Picasso>
    )

    expect(container).toMatchSnapshot()
  })

  test('wrong range', () => {
    let from: number | null, to: number | null

    const errorMessage = () =>
      `Invalid range. Please check the values you have passed: from: ${from}, to: ${to}`

    // casting is needed to trick typescript and pass null
    const tryRender = () =>
      render(
        <Picasso loadFonts={false}>
          <YearSelect
            onChange={() => {}}
            from={from as number}
            to={to as number}
          />
        </Picasso>
      )

    from = null
    expect(tryRender).toThrow(errorMessage())

    to = null
    expect(tryRender).toThrow(errorMessage())

    from = 11
    to = 10
    expect(tryRender).toThrow(errorMessage())
  })
})
