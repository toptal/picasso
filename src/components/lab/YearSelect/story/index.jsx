import PicassoBook from '~/.storybook/components/PicassoBook'

import { YearSelect } from '../YearSelect'

const page = PicassoBook.createPage('YearSelect', `Year select`, 'Lab')

page
  .createTabChapter('Props')
  .addComponentDocs({ component: YearSelect, name: 'YearSelect' })

page
  .createChapter()
  .addExample('lab/YearSelect/story/Default.example.jsx', 'Default')
