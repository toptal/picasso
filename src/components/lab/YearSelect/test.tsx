import React from 'react'
import { render, cleanup } from 'react-testing-library'

import Picasso from '../../Picasso'
import YearSelect from './YearSelect'

afterEach(cleanup)

describe('YearSelect', () => {
  test('default render', () => {
    const { container } = render(
      <Picasso loadFonts={false}>
        <YearSelect from='2001' to='2005' onChange={() => {}} />
      </Picasso>
    )

    expect(container).toMatchSnapshot()
  })
})
