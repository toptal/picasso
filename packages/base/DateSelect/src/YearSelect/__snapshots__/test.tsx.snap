// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearSelect render in descending order 1`] = `
<ul
  class="relative list-none outline-none shadow-2 py-2 px-0 m-0 rounded-sm PicassoScrollMenu-menu PicassoSelectOptions-menu"
  role="listbox"
  tabindex="-1"
>
  <div
    tabindex="0"
  >
    <div
      class="PicassoScrollMenu-scrollView"
    >
      <li
        aria-disabled="false"
        aria-selected="false"
        class="text-black bg-blue hover:bg-blue focus:bg-blue px-4 py-[0.375rem] min-w w-auto min-h md:min-h relative cursor-pointer transition-colors duration-150 ease-in overflow-hidden whitespace-normal text-left no-underline flex items-center justify-start outline-none appearance-none"
        data-highlighted="true"
        role="option"
        tabindex="-1"
        value="2005"
      >
        <div
          class="flex flex-1 max-w"
        >
          <div
            class="flex flex-col flex-1 min-w"
          >
            <div
              class="flex items-center"
            >
              <div
                class="w-4 inline-flex mr-2"
              />
              <span
                class="flex-1 text-md leading-5"
              >
                2005
              </span>
            </div>
          </div>
        </div>
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="text-black bg-[transparent] hover:bg-blue focus:bg-blue px-4 py-[0.375rem] min-w w-auto min-h md:min-h relative cursor-pointer transition-colors duration-150 ease-in overflow-hidden whitespace-normal text-left no-underline flex items-center justify-start outline-none appearance-none"
        data-highlighted="false"
        role="option"
        tabindex="-1"
        value="2004"
      >
        <div
          class="flex flex-1 max-w"
        >
          <div
            class="flex flex-col flex-1 min-w"
          >
            <div
              class="flex items-center"
            >
              <div
                class="w-4 inline-flex mr-2"
              />
              <span
                class="flex-1 text-md leading-5"
              >
                2004
              </span>
            </div>
          </div>
        </div>
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="text-black bg-[transparent] hover:bg-blue focus:bg-blue px-4 py-[0.375rem] min-w w-auto min-h md:min-h relative cursor-pointer transition-colors duration-150 ease-in overflow-hidden whitespace-normal text-left no-underline flex items-center justify-start outline-none appearance-none"
        data-highlighted="false"
        role="option"
        tabindex="-1"
        value="2003"
      >
        <div
          class="flex flex-1 max-w"
        >
          <div
            class="flex flex-col flex-1 min-w"
          >
            <div
              class="flex items-center"
            >
              <div
                class="w-4 inline-flex mr-2"
              />
              <span
                class="flex-1 text-md leading-5"
              >
                2003
              </span>
            </div>
          </div>
        </div>
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="text-black bg-[transparent] hover:bg-blue focus:bg-blue px-4 py-[0.375rem] min-w w-auto min-h md:min-h relative cursor-pointer transition-colors duration-150 ease-in overflow-hidden whitespace-normal text-left no-underline flex items-center justify-start outline-none appearance-none"
        data-highlighted="false"
        role="option"
        tabindex="-1"
        value="2002"
      >
        <div
          class="flex flex-1 max-w"
        >
          <div
            class="flex flex-col flex-1 min-w"
          >
            <div
              class="flex items-center"
            >
              <div
                class="w-4 inline-flex mr-2"
              />
              <span
                class="flex-1 text-md leading-5"
              >
                2002
              </span>
            </div>
          </div>
        </div>
      </li>
      <li
        aria-disabled="false"
        aria-selected="false"
        class="text-black bg-[transparent] hover:bg-blue focus:bg-blue px-4 py-[0.375rem] min-w w-auto min-h md:min-h relative cursor-pointer transition-colors duration-150 ease-in overflow-hidden whitespace-normal text-left no-underline flex items-center justify-start outline-none appearance-none"
        data-highlighted="false"
        role="option"
        tabindex="-1"
        value="2001"
      >
        <div
          class="flex flex-1 max-w"
        >
          <div
            class="flex flex-col flex-1 min-w"
          >
            <div
              class="flex items-center"
            >
              <div
                class="w-4 inline-flex mr-2"
              />
              <span
                class="flex-1 text-md leading-5"
              >
                2001
              </span>
            </div>
          </div>
        </div>
      </li>
    </div>
  </div>
</ul>
`;

exports[`YearSelect renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="makeStyles-root makeStyles-rootFull"
    >
      <div
        class="makeStyles-inputWrapper"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root PicassoOutlinedInput-root PicassoOutlinedInput-rootFull PicassoOutlinedInput-rootMedium makeStyles-outlinedInput MuiInputBase-fullWidth"
          role="textbox"
        >
          <input
            aria-invalid="false"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input PicassoOutlinedInput-input PicassoOutlinedInput-inputMedium"
            readonly=""
            size="1"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline PicassoOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend"
              style="width: 0.01px;"
            >
              <span>
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
        <svg
          class="PicassoSvgDropdownArrows16-root makeStyles-caret"
          style="min-width: 16px; min-height: 16px;"
          viewBox="0 0 16 16"
        >
          <path
            d="m10.997 10.29.707.707-3.707 3.707-3.707-3.707.707-.707 3 3 3-3Zm-3-9 3.707 3.707-.707.707-3-3-3 3-.707-.707 3-3 .707-.707Z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
